arbor/backends/event_stream_base.hpp

```diff
@@ -94,16 +98,75 @@ struct event_stream_base {
                 if (step >= n_steps) break;
                 arb_assert(div + target < handles.size());
                 const auto& handle = handles[div + target];
-                streams[handle.mech_id].ev_data_.push_back({handle.mech_index, weight});
-                dt_sizes[handle.mech_id][step]++;
+                auto& stream = streams[handle.mech_id];
+                stream.spikes_.push_back(spike_data{step, time, handle.mech_index, weight});
+                // insertion sort with last element as pivot
+                auto first = stream.spikes_.begin();
+                auto last = stream.spikes_.end();
+                auto pivot = std::prev(last, 1);
+                std::rotate(std::upper_bound(first, pivot, *pivot, [](auto const& l, auto const& r) noexcept {
+                    if (l.step < r.step) return true;
+                    if (l.step > r.step) return false;
+                    if (l.mech_index < r.mech_index) return true;
+                    if (l.mech_index > r.mech_index) return false;
+                    return l.time < r.time; }),
+                    pivot, last);
+                // increment count in current time interval
+                stream.spike_counter_[step]++;
             }
         }
 
         for (auto& [id, stream]: streams) {
-            util::make_partition(stream.ev_spans_, dt_sizes[id]);
-            stream.init();
+            // copy temporary deliverable_events into stream's ev_data_
+            stream.ev_data_.reserve(stream.spikes_.size());
+            std::transform(stream.spikes_.begin(), stream.spikes_.end(), std::back_inserter(stream.ev_data_),
+                [](auto const& e) noexcept -> arb_deliverable_event_data {
+                return {e.mech_index, e.weight}; });
+            // scan over spike_counter_ and  written to ev_spans_
+            util::make_partition(stream.ev_spans_, stream.spike_counter_);
+            // delegate to derived class init: static cast necessary to access protected init()
+            static_cast<spike_event_stream_base&>(stream).init();
         }
     }
+
+  protected: // members
+    struct spike_data {
+        arb_size_type step = 0;
+        time_type time = 0;
+        cell_local_size_type mech_index = 0;
+        float weight = 0;
```

Comment:
```suggestion
        float weight = 0;
        auto operator<=>(const spike_data&) const = default;
```

Reply:
absolutely, thanks for pointing this out